src/analysis/processing/qgsalgorithmcheckgeometrydangle.cpp

```diff
@@ -0,0 +1,290 @@
+  }
+}
+
+auto QgsGeometryCheckDangleAlgorithm::prepareAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback * ) -> bool
+{
+  mTolerance = parameterAsInt( parameters, QStringLiteral( "TOLERANCE" ), context );
+
+  return true;
+}
+
+auto QgsGeometryCheckDangleAlgorithm::createFeaturePool( QgsVectorLayer *layer, bool selectedOnly ) const -> QgsFeaturePool *
+{
+
+  return new QgsVectorDataProviderFeaturePool( layer, selectedOnly );
+}
+
+static auto outputFields( ) -> QgsFields
+{
+  QgsFields fields;
+  fields.append( QgsField( QStringLiteral( "gc_layerid" ), QVariant::String ) );
+  fields.append( QgsField( QStringLiteral( "gc_layername" ), QVariant::String ) );
+  fields.append( QgsField( QStringLiteral( "gc_featid" ), QVariant::Int ) );
+  fields.append( QgsField( QStringLiteral( "gc_partidx" ), QVariant::Int ) );
+  fields.append( QgsField( QStringLiteral( "gc_ringidx" ), QVariant::Int ) );
+  fields.append( QgsField( QStringLiteral( "gc_vertidx" ), QVariant::Int ) );
+  fields.append( QgsField( QStringLiteral( "gc_errorx" ), QVariant::Double ) );
+  fields.append( QgsField( QStringLiteral( "gc_errory" ), QVariant::Double ) );
+  fields.append( QgsField( QStringLiteral( "gc_error" ), QVariant::String ) );
+  return fields;
+}
+
+
+auto QgsGeometryCheckDangleAlgorithm::processAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback *feedback ) -> QVariantMap
+{
+  QString dest_output;
+  QString dest_errors;
+  std::unique_ptr< QgsProcessingFeatureSource > source( parameterAsSource( parameters, QStringLiteral( "INPUT" ), context ) );
+  if ( !source )
+    throw QgsProcessingException( invalidSourceError( parameters, QStringLiteral( "INPUT" ) ) );
+
+  mInputLayer.reset( source->materialize( QgsFeatureRequest() ) );
+
+  if ( !mInputLayer.get() )
+    throw QgsProcessingException( QObject::tr( "Could not load source layer for INPUT" ) );
+
+  QgsFields fields = outputFields();
+
+  std::unique_ptr< QgsFeatureSink > sink_output( parameterAsSink( parameters, QStringLiteral( "OUTPUT" ), context, dest_output, fields, mInputLayer->wkbType(), mInputLayer->sourceCrs() ) );
+  if ( !sink_output )
+  {
+    throw QgsProcessingException( invalidSinkError( parameters, QStringLiteral( "OUTPUT" ) ) );
+  }
+  std::unique_ptr< QgsFeatureSink > sink_errors( parameterAsSink( parameters, QStringLiteral( "ERRORS" ), context, dest_errors, fields, Qgis::WkbType::Point, mInputLayer->sourceCrs() ) );
+  if ( !sink_errors )
+  {
+    throw QgsProcessingException( invalidSinkError( parameters, QStringLiteral( "ERRORS" ) ) );
+  }
+
+  int resolutionMethod{-1};
+  if ( mIsInPlace )
+  {
+    resolutionMethod = parameterAsEnum( parameters, QStringLiteral( "RESOLUTIONMETHOD" ), context );
+  }
+
+  QgsProcessingMultiStepFeedback multiStepFeedback( mIsInPlace ? 4 : 3, feedback );
+
+  QgsProject *project = mInputLayer->project() ? mInputLayer->project() : QgsProject::instance();
```

Comment:
Can we avoid accessing project here? It's likely a cause of the multithreading issues

Reply:
Backing up a step, why do the geometry checker classes need access to a project at all?

Reply:
"Project can be used to resolve additional layers on algorithms".  Since, it's the layerId that is passed and one has to retrieve the mapLayer from this ID.
AFAIK, it's used in Gap Algorithm Check. Maybe @m-kuhn has a better idea than me on how and where it's used.